### ISSUE TITLE
Question about Homebrew

### ISSUE THREAD

**Issue by michaelmyc** [jasongin/nvs#234]
Hi @jasongin , I've been using your nvs tool and I really liked it. I got a new mac and wanted to use nvs with it, but nvs is not on Homebrew. I quickly searched through the github issues and it seems like you are interested in getting it to work with Homebrew as well but didn't get around to doing it. So I decided to see if I can beat you to it and started working on getting a Homebrew formulae. 

For nvs to work well in a package manager, we need to ensure that upgrading the package doesn't ruin users' Node environments, so I'm considering using soft links to create symlinks of crucial files in `$HOME/.nvs`. 

Thus I'm wondering what files in the repo are crucial to be placed in `$NVS_HOME` for nvs to run on MacOS? From my testing it seems like as long as I have `nvm.sh`, `defaults.json`, `lib`, and `deps` I can use nvs. But I can't be sure if these are sufficient for all the functionalities to be satisfied.

**Comment by jasongin** [comment]
Thanks, I don't use Mac or Homebrew very much so I don't have the time or motivation to maintain a package there.

It should be possible to have `NVS_HOME` point to a directory that does not include any of the script code. That directory would just hold the actual downloaded Node.js files and `settings.json`, while the brew-installed nvs script files (including `lib`, `deps`, and `defaults.json`) live in some other (versioned) location. I have not tested that configuration in a while though.

In `nvs.sh` you'll see `NVS_ROOT` refers to the directory containing the script, while `NVS_HOME` is the directory containing the downloaded Node.js files. Generally there is no assumption that those 2 locations are the same. The `install` command is an exception... but I guess a brew package would use a different install process.

**Comment by ljharb** [comment]
fwiw, i refuse to support homebrew on nvm, because the way homebrew is designed, upgrading nvm via brew wipes out the user's installed nodes, which is a far worse experience imo than telling them to uninstall it with brew and reinstall it properly.

I'm not sure if nvs would have the same problem.

**Comment by jasongin** [comment]
I think we can avoid that problem by keeping `NVS_HOME` directory separate from the installed scripts directory.

**Comment by michaelmyc** [comment]
@jasongin I'm assuming the installation should look like this:
```sh
export NVS_ROOT="SOME_DIR"
git clone https://github.com/jasongin/nvs "$NVS_ROOT"
. "$NVS_ROOT/nvs.sh" install
```
and the shell profile should look something like this:
```sh
export NVS_HOME="$HOME/.nvs"
export NVS_ROOT="SOME_DIR"
[ -s "$NVS_ROOT/nvs.sh" ] && \. "$NVS_ROOT/nvs.sh"
```

> while the brew-installed nvs script files (including `lib`, `deps`, and `defaults.json`) live in some other (versioned) location

So I can assume that these are all the files that are needed for nvs to work?

**Comment by jasongin** [comment]
> I can assume that these are all the files that are needed for nvs to work?

Also add `package.json`, as it is loaded when self-reporting the version. Then that should work. Please test. :)

Also it's not necessary to export `NVS_ROOT` as it is not needed outside the context of those scripts.

**Comment by michaelmyc** [comment]
I'm close to getting an install script working. The manual install method above is working wonders for me and I have a homebrew script copying the actions, however I need to figure out why I'm hitting `Failed to setup node binary.` when downloading node. 

Do we want to officially support Homebrew as an install option or treat it as a community-supported option like in nvm? I think I can manage to maintain it if I can manage to create an install script.

**Comment by jasongin** [comment]
> Do we want to officially support Homebrew as an install option or treat it as a community-supported option like in nvm? I think I can manage to maintain it if I can manage to create an install script.

From my point of view `nvs` can be supported in the same way regardless of whether it was installed with homebrew or other documented means, assuming the homebrew installation is done in a way that I approve. But I personally don't want to support homebrew-specific issues with the installer since I don't have any experience in that area.

One other concern I have is about the release process. When a new version of `nvs` is ready to release, what is the process of publishing it to homebrew? And who has permission to do that?

**Comment by michaelmyc** [comment]
> But I personally don't want to support homebrew-specific issues with the installer since I don't have any experience in that area.

I should be able to help answer some of those questions since I wrote the script. Or I can open a repo specifically for Homebrew-related questions so you can direct people there in the README. 

> When a new version of `nvs` is ready to release, what is the process of publishing it to homebrew? And who has permission to do that?

I think anyone can make modifications and open a PR on [homebrew/core](https://github.com/Homebrew/homebrew-core), and it's up to Homebrew's reviewer to decide if it gets accepted or not. Unless something significant needs to change, it's as simple as copying over the URL to the tarball of the source of the new release and updating the checksum. 

Also, there is testing built into the formulae (install script in Homebrew vocabulary), and we can work on figuring out what tests are appropriate. I'm pretty sure it'll be a big red flag for Homebrew reviewers when you change a formulae's tests. 

If you're not comfortable having it in homebrew/core (more discoverable), we can also serve it as a 3rd party [Tap](https://docs.brew.sh/Taps) so we have full control of the repo and how the script gets updated.

**Comment by jasongin** [comment]
> Or I can open a repo specifically for Homebrew-related questions

I think there's no need for a separate repo just for questions, unless we also wanted to use it for a tap.

> If you're not comfortable having it in homebrew/core (more discoverable), we can also serve it as a 3rd party Tap.

Putting it in core is probably fine.

**Comment by jasongin** [comment]
> I should be able to help answer some of those questions since I wrote the script.

I'm assuming/hoping once it's set up then there shouldn't be many issues/questions, as it is not a complicated installation.

**Comment by michaelmyc** [comment]
I don't think it should cause a lot of problems either. But if there are any problems/requests. 

Also, it seems like nvs on chocolatey is still on v1.5.2 which was released way back in 2018. I think it's a good idea for someone to update it.

**Comment by michaelmyc** [comment]
I'm really close to getting a formulae working. It seems like if I use the 1.6.0 release tarball is broken on M1 Macs, and since I'm using an M1 Mac, I wouldn't be able to test the formulae. 😢

We should release a 1.6.1 version that supports M1 Macs.

**Comment by michaelmyc** [comment]
@jasongin PR #222 might be interesting to merge for natively support of M1 Macs (better performance than transpiled x64 versions) in the new release.

**Comment by jasongin** [comment]
See my comment in #222 - I have some hesitation about merging that PR as it is.

But regardless of what we choose to do there, I will do a 1.6.1 release first that unblocks M1 by using x64 builds. Apparently I forgot to do that before.

**Comment by jasongin** [comment]
https://github.com/jasongin/nvs/releases/tag/v1.6.1

**Comment by michaelmyc** [comment]
Opened a PR in Homebrew. https://github.com/Homebrew/homebrew-core/pull/98776

**Comment by michaelmyc** [comment]
>> Another caveat is that the package will install a bootstrapping Node.js in $NVS_HOME if it isn't already installed. This is unavoidable behavior and cannot be addressed with adding node to dependencies.
>
> That conflicts with https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae unfortunately. If this can't be disabled we'll have to reject it.

@jasongin Hmm. Is it possible we add a command line option or environment variable to use the system's Node.js to run nvs? Or else we can just publish it as a 3rd party Tap now that I got the formula working. 

I can look into making the code changes if you're swamped on things.

**Comment by jasongin** [comment]
Maybe possible, though I have a few concerns:
 - It may seem a little funny (or confusing?) if the tool for installing Node.js versions does a system install of node.js.
 - Would the `nvs` brew package depend on the node.js brew package?
     - I'm assuming `brew` would block uninstalling the `node` package when the `nvs` package depends on it?
 - It should require a specific version (at least major version) of Node.js. I don't test `nvs` with bootstrap node versions different from the configured version. I think it requires at least v10 now. While future node major versions _should_ be compatible, again that is not normally tested.
 - What about privileges? Installing a system Node.js normally requires root access, whereas setting up `nvs` (including the boostrap node version that it downloads) does not.

**Comment by michaelmyc** [comment]
Homebrew can specify major Node.js versions like `node@12`. And I agree with you that this doesn't make a ton of sense in the context of nvs. We are installing a tool to manage Node while depending on a Node.js install the tool cannot manage. 

But I think they've passed `miniconda` and `anaconda` as a Cask, which are basically the Python equivalent of `nvs`. Maybe we can try to go and try that out based on this in their docs: 

![image](https://user-images.githubusercontent.com/15188392/162540345-4deb293f-b771-451b-8350-166e00553ca0.png)

**Comment by michaelmyc** [comment]
@jasongin I opened a new [PR](#236) that includes a homebrew install script. We'll need a new release once we check that in.

**Comment by michaelmyc** [comment]
@jasongin Realized a patch is needed for Homebrew upgrades to function correctly: #237
